### PR TITLE
API proxy authenticator: relax key set config validation, wip

### PR DESCRIPTION
**Pull request by jgehrcke**
This addresses https://github.com/opstrace/opstrace/issues/450.

### CODE REVIEW

File: go/pkg/authenticator/keyset_test.go [hunk 1]
```diff
@@ -44,3 +44,14 @@ func TestKeysetFromEnv_TwoKeys(t *testing.T) {
...
+
+func TestKeysetFromEnv_EmtpySetButFallback(t *testing.T) {
+	os.Setenv("API_AUTHTOKEN_VERIFICATION_PUBKEY_SET", "")
+	os.Setenv("API_AUTHTOKEN_VERIFICATION_PUBKEY", TestPubKey)
+
+	// This is now expected to _not_ crash, becuse a fallback key is
+	// configured.
+	ReadConfigFromEnvOrCrash()
+	log.Infof("keyset map:\n%v", authtokenVerificationPubKeys)
+	log.Infof("fallback key:\n%v", authtokenVerificationPubKeyFallback)
```
Comment (jgehrcke) [inline go/pkg/authenticator/keyset_test.go:56]: this tests the scenario that led to the #450 breakage.

**Comment by jgehrcke** [comment]
`main-test-gcp` failed with #151: https://github.com/opstrace/opstrace/issues/151#issuecomment-792301414

**Comment by sreis** [comment]
@jgehrcke If I download the [CLI built from this PR](https://buildkite.com/opstrace/prs/builds/3736#12c5fd27-c744-48d1-af51-005036af887d/208-263) and run the upgrade command would that be enough for the controller to deploy the APIs with the correct env variables? Or do we need to do something else?

**Comment by jgehrcke** [comment]
> would that be enough for the controller to deploy the APIs with the correct env variables?

Hm, generally -- as I haven't tested this myself, let's assume things are still broken!

The new API proxy versions are supposed to work in view of the _old_ environment variable being set.

The state is:

- _old_ controller config map
- new controller
- new API proxy versions

When I think about it that way, things are probably still broken. I was thinking of the following state when I built this patch here:

- old controller config map
- old controller
- new API proxy versions

Woof.

This is a great example of complexity that's certainly possible to manage, and even rather easy to manage buuut... where in practice things are just so so likely to break.

**Comment by jgehrcke** [comment]
What one could do, hypothetically, is to make the newer controller version work with an older controller config, transparently.

However, that can't quite be maintained.

What needs to be done is to migrate the controller config to the new schema, I think.

1) Read the current controller config map.
2) Parse current authenticator public key from it
3) Construct new authenticator key set configuration document (JSON), which involves building a key ID, following these instructions: https://github.com/opstrace/opstrace/tree/main/go/pkg/authenticator#key-id-calculation
4) Building a new controller config object, with the parameter `tenant_api_authenticator_pubkey_set_json`.
5) Overwrite the controller config map.

But where? Should the controller itself do this? Sounds like the right thing to do: the controller should perform migrations for its own config, upon startup. What do we think?

----

We could also use this opportunity and re-evaluate the current upgrade strategy (that we're currently trying to make successful). I know it's a potentially hard thing to think about -- but maybe we really do want to use the more easy-to-reason-about hibernation upgrade strategy where a completely new cluster is being used against the 'old' DB and cloud storage.

**Pull request merged**